ome/datasource_firmware_catalog_test.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright (c) 2023 Dell Inc., or its subsidiaries. All Rights Reserved.
+Licensed under the Mozilla Public License Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://mozilla.org/MPL/2.0/
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ome
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"terraform-provider-ome/helper"
+	"terraform-provider-ome/utils"
+	"testing"
+
+	. "github.com/bytedance/mockey"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestDataSource_FirmwareCatalogRead(t *testing.T) {
+	if os.Getenv("TF_ACC") == "" {
+		t.Skip("Dont run with units tests because it will try to create the context")
+	}
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:                 func() { testAccPreCheck(t) },
+		ProtoV6ProviderFactories: testAccProtoV6ProviderFactories,
+		Steps: []resource.TestStep{
+			// No filter so should return all values
+			{
+				Config: firmwareCatAll + outputs,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckOutput("fetched_any", "true"),
+					resource.TestCheckOutput("fetched_many", "true"),
+					resource.TestCheckOutput("fetched_one", "false"),
+				),
+			},
+			// Empty Filter so should return all values
+			{
+				Config: firmwareCatAllEmptyNameFilter + outputs,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckOutput("fetched_any", "true"),
+					resource.TestCheckOutput("fetched_many", "true"),
+					resource.TestCheckOutput("fetched_one", "false"),
+				),
+			},
+			// Using the name filter
+			{
+				Config: firmwareCatFilter + outputs,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckOutput("fetched_any", "true"),
+					resource.TestCheckOutput("fetched_many", "false"),
+					resource.TestCheckOutput("fetched_one", "true"),
+				),
+			},
+			// Error getting firmware catalogs
+			{
+				PreConfig: func() {
+					FunctionMocker = Mock(helper.GetAllCatalogFirmware).Return(nil, fmt.Errorf("Mock error")).Build()
+				},
+				Config:      firmwareCatAll,
+				ExpectError: regexp.MustCompile(`.*Error fetching firmware catalogs*.`),
+			},
+			// Processing error test
+			{
+				PreConfig: func() {
+					if FunctionMocker != nil {
+						FunctionMocker.UnPatch()
+					}
+					FunctionMocker = Mock(utils.CopyFields).Return(fmt.Errorf("Mock error")).Build()
+				},
+				Config:      firmwareCatAll,
+				ExpectError: regexp.MustCompile(`.*Error processing firmware catalogs*.`),
+			},
+		},
+	})
+}
+
+var outputs = `
+output "fetched_many" {
+	value = length(data.ome_firmware_catalog.data-catalog.firmware_catalogs) > 1
```

Comment:
@doriac11 This could return false if there's only one resource on array, so may be check _!=0_ here ?

Reply:
There is a check for a single catalog, and multiple catalogs, so in this test case we want it to be greater then one. if you use != 0 then 1 would be a valid condition which we dont want for this case.

Reply:
@doriac11 Yes this should ideally be a test for Get All catalog API(or firmwareCatAll config) so if there is only one catalog on the array, then also it should not fail. For acceptance testing if we run the test and the array has only one catalog wouldn't this test will fail (when ideally it isn't a fail).